### PR TITLE
Win32 command line quoting/escaping

### LINKED ISSUE #65: Windows command line generation doesn't handle some edge cases
**Describe the bug**
While creating tests for my Windows backup scripts I found a few cases that were not handled correctly by `startProcess`:
* Argument array `['cacls.exe', 'C:\\']` is converted to command line `"cacls.exe" "C:\"` where the trailing `\"` is treated as an inline double quote.  Therefore the argument that `cacls` actually sees is `C:"`, which is not what was intended.
* The double quotes that are added around every argument whether needed or not confuse some tools, e.g. argument array `['subst.exe', '/D', 'Z:']` is converted to command line `"subst.exe" "/D" "Z:"`.  The quotes around the `/D` mean that `subst` doesn't recognise the argument... arguably that's its problem not ours, but as these quotes are unnecessary and unavoidable with the current implementation it's hard to work around it.

Going down the rabbit hole of how Windows command line parsers really work turned up a number of other weird rules that won't always be followed by the current implementation.

The linked PR changes the implementation of the Win32 process helper to more closely follow the rules for Windows command lines.

**To Reproduce**
Many (not all) of the argument lists in test `PySys_internal_186` from the linked PR will fail or give undefined results with the current implementation.  For example:
* `self.startProcess('cacls.exe', ['C:\\'], environs=os.environ, stdouterr="cacls")` will fail with a `There are no more files.` error.
* `self.startProcess('subst.exe', ['/D', 'Z:'], environs=os.environ, stdouterr="subst")` will fail with an `Invalid parameter - "/D"` error.

**Expected behavior**
Generated windows command lines should follow the documented (and undocumented) rules used by the standard Windows command line parsers, `parse_cmdline` in the VS CRT library and `CommandLineToArgvW` in the Win32 API.  See these links for a lot more detail on what the rules actually are:
* https://docs.microsoft.com/en-us/previous-versions/17w5ykft(v=vs.85)
* http://www.windowsinspired.com/the-correct-way-to-quote-command-line-arguments/
* http://www.windowsinspired.com/understanding-the-command-line-string-and-arguments-received-by-a-windows-program/
* http://www.windowsinspired.com/how-a-windows-programs-splits-its-command-line-into-individual-arguments/
* https://daviddeley.com/autohotkey/parameters/parameters.htm

**Affected version(s)**
Observed with PySys 1.6 and 2.1 so I assume all versions are affected.
Not Python version-specific.
Obviously Windows-only.

### PR DESCRIPTION
**Pull request by rsm31**
Improve quoting/escaping of generated Windows command lines.  The current implementation doesn't handle some edge cases correctly, for example:

* Trailing backslashes: An argument `C:\` will be quoted as `"C:\"` and then incorrectly parsed as `C:"` - the `\"` is treated as a literal quote, not an argument delimiter.
* Inline backslashes followed by double quotes: Very much an edge case but one that needs to be handled by escaping the backslashes and the double quote.  Annoyingly, backslashes *not* followed by a double quote don't need to be escaped!
* Inline double quotes: The approach of repeating an inline double quote character so that it is parsed as a literal quote rather than as an argument delimiter works most of the time, but it is handled slightly differently by the two standard command line parsers so results with some command lines will be undefined.
* Unnecessary quotes: Every argument is wrapped in double quotes whether needed or not, which confuses some legacy tools with apparently very dumb command line parsing.
* The rules for the command part of the command line (i.e. `argv[0]`) are different from the arguments, and quite different between the two standard parsers.  Fortunately most of the differences are in "don't do that" situations where the command wouldn't be valid anyway, so most cases can be handled by wrapping quotes around a command path that contains spaces.  The other quoting and escaping rules used for arguments do *not* apply to the command.

This PR changes Windows command line generation to handle or avoid all these cases and adds a test to verify that generated command lines are parsed correctly by `CommandLineToArgvW` which is accessible through the `pywin32` API.  "Correct" means that we get back the same argument array that we started with.  It might be overkill (it certainly wasn't necessary to go this far to fix the specific issue I was having) but once down the rabbit hole it seemed like a good idea to try to fix everything...

All existing tests continue to pass after this change.

Fixes #65.

### CODE REVIEW

**Comment by rsm31** [comment]
Oops, should have tested on Linux too - stand by.

File: pysys/process/plat-win32/helper.py [hunk 1]
```diff
@@ -112,11 +112,79 @@ def _writeStdin(self, data):
...
+		https://docs.microsoft.com/en-us/previous-versions/17w5ykft(v=vs.85)
+		http://www.windowsinspired.com/the-correct-way-to-quote-command-line-arguments/
+		http://www.windowsinspired.com/understanding-the-command-line-string-and-arguments-received-by-a-windows-program/
+		http://www.windowsinspired.com/how-a-windows-programs-splits-its-command-line-into-individual-arguments/
+		https://daviddeley.com/autohotkey/parameters/parameters.htm
+		This method tries to avoid any areas of different behaviour between the two
+		standard parsers, in particular the undocumented rules around handling of
+		consecutive unescaped double quote characters.
+		"""
+		output = ""
```
Comment (ben-spiller) [inline pysys/process/plat-win32/helper.py:150]: This _could_ be premature optimization, but we don't have the data to be sure and this startProcess code is executed a LOT... so how about reverting to a trivial but safe implementation if input doesn't contain space/tab/doublequote chars (or backslash??)?

**Review by ben-spiller — CHANGES_REQUESTED** [review]
Presumably https://docs.python.org/3.9/library/shlex.html doesn't take care of windows escaping? (from the docs I wasn't 100% sure but probably not)

File: test/correctness/PySys_internal_186/pysystest.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		2. CommandLineToArgv() on the command line gets back the original arguments
+		"""
+		newcmd, cmdline = self.pimpl._ProcessImpl__buildCommandLine(command, args)
+		oldargv = [command] + args
+		newargv = win32api.CommandLineToArgv(cmdline)
+		self.log.info("-----")
+		self.log.info("Original: %s" % oldargv)
+		self.log.info("Quoted:   %s" % cmdline)
+		self.log.info("Parsed:   %s" % newargv)
+		if expected and expected != cmdline:
```
Comment (ben-spiller) [inline test/correctness/PySys_internal_186/pysystest.py:44]: slightly gross - why not use assertThat? This will automatically give nice (multi-line) comparison messages if there's a mismatch

**Review by ben-spiller — CHANGES_REQUESTED** [review]

File: test/correctness/PySys_internal_186/pysystest.py (hunk 2 above)
Comment (rsm31) [reply test/correctness/PySys_internal_186/pysystest.py:44]: You're right, that is quite disgusting.  Fixed.

File: .github/workflows/pysys-test.yml [hunk 3]
```diff
@@ -192,7 +192,9 @@ jobs:
            
            # must do this AFTER testing to avoid polluting the installed packages while running the tests
            # peg to specific versions to avoid our build breaking when there's an upgrade
-           pip install docutils==0.16 sphinx==3.0.1 sphinx-epytext==0.0.4 sphinx_rtd_theme==0.4.3
+           echo TEMPORARILY disabling PYTHONWARNINGS until we can move to Jinja2>=3.1
```
[the remark this answers is not in the document]
Comment (rsm31) [reply .github/workflows/pysys-test.yml:195]: Jinja2 is indeed now pinned to v3.0.3.  That version produces a deprecation warning about something called `environmentfilter` going away - and in v3.1 it's gone.  Probably could also resolve by bumping the version of sphinx and other stuff, but this seemed like the quickest fix given that I don't actually know what most of these packages do :-)

File: pysys/process/plat-win32/helper.py (hunk 1 above)
Comment (rsm31) [reply pysys/process/plat-win32/helper.py:150]: Hmm... I would expect the cost of actually starting the process to dominate (I remember that being relatively expensive on Windows) but I can whip up a quick performance test to compare with the old implementation.  If it's worthwhile then falling back to a dumb implementation for the simple (and likely most common) case seems reasonable.  Stay tuned.

**Comment by rsm31** [comment]
> Presumably https://docs.python.org/3.9/library/shlex.html doesn't take care of windows escaping? (from the docs I wasn't 100% sure but probably not)

Pretty sure it only handles Unix/POSIX-style shells and quoting.
A command path with spaces is handled like so:
```
>>> import shlex
>>> shlex.join(['C:\\path with spaces\\program.exe', 'arg'])
"'C:\\path with spaces\\program.exe' arg"
```
and if I try to execute that in PowerShell:
```
(.venv) PS C:\Users\scott\Projects\pysys-test\test> 'C:\\path with spaces\\program.exe' arg
At line:1 char:35
+ 'C:\path with spaces\program.exe' arg
+                                   ~~~
Unexpected token 'arg' in expression or statement.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : UnexpectedToken
```
or Command Prompt:
```
C:\Users\scott>'C:\path with spaces\program.exe' arg
The filename, directory name, or volume label syntax is incorrect.
```
And it definitely won't handle the mad rules about double quotes after backslashes:
```
>>> shlex.join(['C:\\path with spaces\\program.exe', 'arg\\\"'])
'\'C:\\path with spaces\\program.exe\' \'arg\\"\''
```
(the backslash after "arg" should be escaped because it's followed by a double quote...)

File: pysys/process/plat-win32/helper.py (hunk 1 above)
Comment (rsm31) [reply pysys/process/plat-win32/helper.py:150]: Well this was probably a bit unnecessary, but I wrote a performance test (PySys_internal_perf_003) and once I had that it seemed only reasonable to make some optimisations... Anyway, full results below but in summary all the easy cases are now faster than they were, and the not-so-easy ones (basically anything with double-quotes _within_ an element of argv) are significantly slower.  I don't think those are particularly common cases, and even on my ancient laptop I could still generate around 6,000+ command lines/second in the worst case, so it shouldn't be a big problem in any reasonable real-world situation.

Results below are comparing several versions of the new algorithm to the "legacy" implementation so lower speedup is actually better.  "unoptimised" is my first attempt (all round bad performance), "shortcircuit" special-cases the easy stuff, and "stringjoin" uses `join()` instead of string concatenation - the last one shows the best overall improvement and is what's in the PR.  BTW the performance results comparison stuff is really cool :-)
```
Performance comparison summary between the following runs (the differing "run details" are shown):
- 2022-06-12_16.59.40.win.unoptimised (8 result keys, 10 samples/result):
     vcsCommit=c93aa462, startTime=2022-06-12 16:59:40
- 2022-06-12_17.28.28.win.shortcircuit (8 result keys, 10 samples/result):
     vcsCommit=c93aa462, startTime=2022-06-12 17:28:28
- 2022-06-12_22.27.58.win.stringjoin (8 result keys, 10 samples/result):
     vcsCommit=95fc7a39, startTime=2022-06-12 22:27:58
- 2022-06-12_15.51.24.win.legacy (8 result keys, 10 samples/result):
     vcsCommit=c93aa462, xargs=useLegacyAlgorithm-True, startTime=2022-06-12 15:51:24

Comparison results:
  where fromRun->toRun format is: (%improvement) = (speedup ratio) = (sigmas/stdDevs where change above +/- 2.0 gives 95% probability it's significant; only significant results are colored)

Command line generation rate (1: command / no args) from PySys_internal_perf_003
 Mean from this run = 684,124 /s with stdDev=231,645 (33.9% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: -48.2% = 0.5183x speedup = -1.4 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: -37.3% = 0.6268x speedup = -1.1 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -39.9% = 0.6007x speedup = -1.2 sigmas

Command line generation rate (2: command with path / no args) from PySys_internal_perf_003
 Mean from this run = 665,764 /s with stdDev=211,251 (31.7% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: -48.9% = 0.5108x speedup = -1.5 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: -37.2% = 0.6281x speedup = -1.2 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -39.8% = 0.6017x speedup = -1.3 sigmas

Command line generation rate (3: command with spaces / no args) from PySys_internal_perf_003
 Mean from this run = 658,374 /s with stdDev=211,629 (32.1% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: -29.1% = 0.7091x speedup = -0.9 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: -13.0% = 0.8705x speedup = -0.4 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -20.0% = 0.8003x speedup = -0.6 sigmas

Command line generation rate (4: simple args) from PySys_internal_perf_003
 Mean from this run = 79,617 /s with stdDev=25,669 (32.2% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: +488.4% = 5.884x speedup = +15.1 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy:  -7.7% = 0.9234x speedup = -0.2 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -46.6% = 0.5342x speedup = -1.4 sigmas

Command line generation rate (5: args with path) from PySys_internal_perf_003
 Mean from this run = 76,030 /s with stdDev=24,456 (32.2% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: +418.6% = 5.186x speedup = +13.0 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: -11.5% = 0.8848x speedup = -0.4 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -50.0% = 0.4998x speedup = -1.6 sigmas

Command line generation rate (6: args with spaces) from PySys_internal_perf_003
 Mean from this run = 79,786 /s with stdDev=24,010 (30.1% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: +527.5% = 6.275x speedup = +17.5 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: +17.1% = 1.171x speedup = +0.6 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: -29.6% = 0.7039x speedup = -1.0 sigmas

Command line generation rate (7: args with quotes) from PySys_internal_perf_003
 Mean from this run = 66,775 /s with stdDev=18,397 (27.6% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: +408.0% = 5.08x speedup = +14.8 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: +528.2% = 6.282x speedup = +19.2 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: +413.5% = 5.135x speedup = +15.0 sigmas

Command line generation rate (8: complex args) from PySys_internal_perf_003
 Mean from this run = 52,031 /s with stdDev=14,779 (28.4% of mean)
   2022-06-12_16.59.40.win.unoptimised->2022-06-12_15.51.24.win.legacy: +630.2% = 7.302x speedup = +22.2 sigmas
  2022-06-12_17.28.28.win.shortcircuit->2022-06-12_15.51.24.win.legacy: +806.0% = 9.06x speedup = +28.4 sigmas
    2022-06-12_22.27.58.win.stringjoin->2022-06-12_15.51.24.win.legacy: +602.7% = 7.027x speedup = +21.2 sigmas
```

Comment (rsm31) [reply pysys/process/plat-win32/helper.py]: Not sure if you read the previous set of performance results but if not, don't bother.  My laptop was suffering from some thermal problems and throttling the CPU under load, so I wouldn't trust any those numbers.  Anyway that problem is fixed and I have re-run the the performance test.  The runs compared below are:
1. `legacy` - the old, simple, but not correct in all cases algorithm
2. `optimised1` - the new algorithm with some easy/common cases short-circuited and all the string concatenation replaced with `join()`s
3. `optimised2` - small tweaks to make sure we get the string length and scan for whitespace exactly once

In summary, compared to the legacy algorithm:
* Cases with no whitespace and no inline double quotes are a bit faster
* Cases with whitespace but no inline double quotes are a bit slower
* Anything with inline double quotes is a _lot_ slower, but still doing ~9,000 command lines/sec on a relatively ancient laptop in the worst case, so I wouldn't expect it to be a problem in the real world, especially as I don't expect this to be a common use case anyway

What do you think?  My view would be that any performance hit in a realistic scenario is going to be lost in the noise...

```
Performance comparison summary between the following runs (the differing "run details" are shown):
- 2022-06-18_10.09.40.win.legacy (8 result keys, 10 samples/result):
     xargs=useLegacyAlgorithm-True, startTime=2022-06-18 10:09:40
- 2022-06-18_10.24.03.win.optimised1 (8 result keys, 10 samples/result):
     startTime=2022-06-18 10:24:03
- 2022-06-18_10.49.40.win.optimised2 (8 result keys, 10 samples/result):
     startTime=2022-06-18 10:49:40

Comparison results:
  where fromRun->toRun format is: (%improvement) = (speedup ratio) = (sigmas/stdDevs where change above +/- 2.0 gives 95% probability it's significant; only significant results are colored)

Command line generation rate (1: command / no args) from PySys_internal_perf_003
 Mean from this run = 1,344,825 /s with stdDev=118,765 (8.8% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2:  +8.0% = 1.08x speedup = +0.9 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  -1.6% = 0.9838x speedup = -0.2 sigmas

Command line generation rate (2: command with path / no args) from PySys_internal_perf_003
 Mean from this run = 1,308,619 /s with stdDev=37,703 (2.9% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2:  +7.2% = 1.072x speedup = +1.0 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  -0.9% = 0.9914x speedup = -0.3 sigmas

Command line generation rate (3: command with spaces / no args) from PySys_internal_perf_003
 Mean from this run = 996,372 /s with stdDev=17,481 (1.8% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: -17.6% = 0.8239x speedup = -2.0 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  -1.3% = 0.9872x speedup = -0.6 sigmas

Command line generation rate (4: simple args) from PySys_internal_perf_003
 Mean from this run = 174,074 /s with stdDev=2,914 (1.7% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: +24.1% = 1.241x speedup = +4.9 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  -2.4% = 0.9758x speedup = -0.3 sigmas

Command line generation rate (5: args with path) from PySys_internal_perf_003
 Mean from this run = 178,622 /s with stdDev=2,242 (1.3% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: +30.4% = 1.304x speedup = +8.6 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  +8.1% = 1.081x speedup = +0.5 sigmas

Command line generation rate (6: args with spaces) from PySys_internal_perf_003
 Mean from this run = 112,490 /s with stdDev=2,422 (2.2% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: -20.9% = 0.791x speedup = -9.7 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  -0.4% = 0.9963x speedup = -0.0 sigmas

Command line generation rate (7: args with quotes) from PySys_internal_perf_003
 Mean from this run = 15,974 /s with stdDev=170.3 (1.1% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: -86.5% = 0.1355x speedup = -7.1 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  +6.1% = 1.061x speedup = +0.6 sigmas

Command line generation rate (8: complex args) from PySys_internal_perf_003
 Mean from this run = 8,996 /s with stdDev=335.9 (3.7% of mean)
      2022-06-18_10.09.40.win.legacy->2022-06-18_10.49.40.win.optimised2: -90.5% = 0.09485x speedup = -19.0 sigmas
  2022-06-18_10.24.03.win.optimised1->2022-06-18_10.49.40.win.optimised2:  +4.8% = 1.048x speedup = +0.4 sigmas
```

**Comment by rsm31** [comment]
> @ben-spiller Do you understand why the code coverage has gone down in `pysys/internal/initlogging.py`, which I haven't touched?

Hmm, looks like a racy test?  The exception on lines 75-77 of `initlogging.py` sometimes gets hit, sometimes not.  Everything passed on a re-run so I don't plan to investigate further right now.

File: pysys/process/plat-win32/helper.py (hunk 1 above)
Comment (ben-spiller) [reply pysys/process/plat-win32/helper.py]: so it's just a bit slower for the common case? sounds ok then! 

i'm a little more cautious about performance of critical-path stuff (like starting processes) the more CPUs we get in modern machines given the GIL problem. i.e. I don't care about the time taken for a single test, but want to keep scalability good when running lots of tests in parallel

File: .github/workflows/pysys-test.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ben-spiller) [reply .github/workflows/pysys-test.yml:195]: Please back out changes to this file rebase, I've fixed it in master

[the remark this answers is not in the document]
Comment (rsm31) [reply .github/workflows/pysys-test.yml]: Done, but now seeing this in the doc build:
```
2022-11-20T12:10:35.3855989Z     from jinja2 import environmentfilter
2022-11-20T12:10:35.3857252Z ImportError: cannot import name 'environmentfilter' from 'jinja2' (/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/jinja2/__init__.py)
```
... which I think is the sae thing I had hacked around previously.  What did I miss?

**Review by ben-spiller — CHANGES_REQUESTED** [review]
All good, but please add a brief line to CHANGELOG.rst (no need to go into detail of the algorithm but give say what the scope of the changes is)

**Comment by rsm31** [comment]
> All good, but please add a brief line to CHANGELOG.rst (no need to go into detail of the algorithm but give say what the scope of the changes is)

Done, hope that's OK?

**Review by ben-spiller — APPROVED** [review]
